### PR TITLE
add N50 calculation to assembly

### PR DESCRIPTION
**Pull request by tedsharpe**

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java [hunk 1]
```diff
@@ -17,6 +18,25 @@ public FermiLiteAssembly( final List<Contig> contigs ) {
...
+        final int[] lengths = new int[nContigs];
+        for ( int idx = 0; idx != nContigs; ++idx ) {
+            lengths[idx] = contigs.get(idx).getSequence().length;
+        }
+        Arrays.sort(lengths);
+        final int totalSize = Arrays.stream(lengths).sum();
+        int runningTotal = 0;
+        for ( int idx = nContigs - 1; idx >= 0; --idx ) {
+            runningTotal += 2 * lengths[idx];
+            if ( runningTotal >= totalSize ) return lengths[idx];
```
Comment (TedBrookings) [inline src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java]: Total nitpick, but according to the definition I found, N50 is the longest contig length such that 50% or more of the genome is in a contig that large or greater. In which case this should check "runningTotal > totalSize" (counter-example, lengths = {4, 4, 8} should yield 8).

File: src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java [hunk 2]
```diff
@@ -17,6 +18,25 @@ public FermiLiteAssembly( final List<Contig> contigs ) {
...
+        final int nContigs = contigs.size();
+        if ( nContigs < 1 ) return 0;
+        if ( nContigs == 1 ) return contigs.get(0).getSequence().length;
+
+        final int[] lengths = new int[nContigs];
+        for ( int idx = 0; idx != nContigs; ++idx ) {
+            lengths[idx] = contigs.get(idx).getSequence().length;
+        }
+        Arrays.sort(lengths);
+        final int totalSize = Arrays.stream(lengths).sum();
```
Comment (TedBrookings) [inline src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java]: I'm not experienced enough in Java to know for sure (or in this project to know if it matters / how this function will be used), but would it be more performant to add up totalSize in the for loop that gets the lengths?

**Review by TedBrookings — APPROVED** [review]
Two extreme nitpicks you should feel free to take or leave. Otherwise totally fine.

File: src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java (hunk 2 above)
Comment (tedsharpe) [reply src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java]: Yes, that would probably be faster.  I'll change it.

File: src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java (hunk 1 above)
Comment (tedsharpe) [reply src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java]: Hmm.  Looks to me like it's correct.  Did you notice that the indices are running backwards?  (I'll add a comment and some javadoc.)
For your example:
totalSize = 4+4+8 = 16
first time through loop,
idx = 2
lengths[2] = 8
runningTotal = 2 * 8 = 16
runningTotal >= totalSize, so return lengths[2] = 8

**Comment by tedsharpe** [comment]
Better?

File: src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java (hunk 1 above)
Comment (TedBrookings) [reply src/main/java/org/broadinstitute/hellbender/utils/fermi/FermiLiteAssembly.java]: I did *NOT* notice that the indices were running backwards. My mistake, your code is correct.

**Pull request merged**